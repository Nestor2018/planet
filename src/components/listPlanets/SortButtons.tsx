import React from "react";
import { View, Text, TouchableOpacity, StyleSheet } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import colors from "../../utils/colors";

interface Props {
  sortOrder: "asc" | "desc";
  setSortOrder: (order: "asc" | "desc") => void;
}

const SortButtons: React.FC<Props> = ({ sortOrder, setSortOrder }) => {
  return (
    <View style={styles.sortContainer} testID="sort-buttons">
      {["asc", "desc"].map((order) => (
        <TouchableOpacity
          key={order}
          style={[
            styles.sortButton,
            sortOrder === order && styles.activeButton,
          ]}
          testID={`button-${order}`}
          onPress={() => setSortOrder(order as "asc" | "desc")}
        >
          <Ionicons
            name={order === "asc" ? "arrow-up" : "arrow-down"}
            size={18}
            color={sortOrder === order ? colors.blackPearl : colors.white}
          />
          <Text
            style={[
              styles.buttonText,
              sortOrder === order && styles.activeButtonText,
            ]}
          >
            {order === "asc" ? "Ascendente" : "Descendente"}
          </Text>
        </TouchableOpacity>
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  sortContainer: {
    flexDirection: "row",
    justifyContent: "center",
    marginVertical: 10,
  },
  sortButton: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: colors.charade,
    padding: 10,
    borderRadius: 8,
    marginHorizontal: 5,
    minWidth: 120,
    justifyContent: "center",
  },
  activeButton: {
    backgroundColor: colors.picton,
  },
  buttonText: {
    color: colors.white,
    marginLeft: 5,
    fontSize: 14,
  },
  activeButtonText: {
    color: colors.blackPearl,
  },
});

export default SortButtons;
