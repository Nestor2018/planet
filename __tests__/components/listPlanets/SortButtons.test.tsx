import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import SortButtons from "../../../src/components/listPlanets/SortButtons";

jest.mock("@expo/vector-icons", () => ({
  Ionicons: ({ name, size, color, ...props }) => (
    <view
      testID={`icon-${name}`}
      name={name}
      size={size}
      color={color}
      {...props}
    />
  ),
}));

describe("SortButtons Component", () => {
  const mockSetSortOrder = jest.fn();

  beforeEach(() => {
    mockSetSortOrder.mockClear();
  });

  it("renders ascending and descending buttons", () => {
    const { getByTestId, getByText } = render(
      <SortButtons sortOrder="asc" setSortOrder={mockSetSortOrder} />,
    );

    expect(getByTestId("sort-buttons")).toBeTruthy();

    expect(getByText("Ascendente")).toBeTruthy();
    expect(getByText("Descendente")).toBeTruthy();

    const ascIcon = getByTestId("icon-arrow-up");
    const descIcon = getByTestId("icon-arrow-down");
    expect(ascIcon).toBeTruthy();
    expect(descIcon).toBeTruthy();
  });

  it("calls setSortOrder with 'asc' when ascending button is pressed", () => {
    const { getByText } = render(
      <SortButtons sortOrder="desc" setSortOrder={mockSetSortOrder} />,
    );

    const ascButton = getByText("Ascendente");
    fireEvent.press(ascButton);

    expect(mockSetSortOrder).toHaveBeenCalledTimes(1);
    expect(mockSetSortOrder).toHaveBeenCalledWith("asc");
  });

  it("calls setSortOrder with 'desc' when descending button is pressed", () => {
    const { getByText } = render(
      <SortButtons sortOrder="asc" setSortOrder={mockSetSortOrder} />,
    );

    const descButton = getByText("Descendente");
    fireEvent.press(descButton);

    expect(mockSetSortOrder).toHaveBeenCalledTimes(1);
    expect(mockSetSortOrder).toHaveBeenCalledWith("desc");
  });

  // 3. Estilos condicionales
  it("applies active styles to ascending button when sortOrder is 'asc'", () => {
    const { getByText, getByTestId } = render(
      <SortButtons sortOrder="asc" setSortOrder={mockSetSortOrder} />,
    );

    const ascButton = getByText("Ascendente").parent;
    const ascIcon = getByTestId("icon-arrow-up");
    const ascText = getByText("Ascendente");

    expect(ascButton.props.style).toStrictEqual([
      { color: "#fff", fontSize: 14, marginLeft: 5 },
      { color: "#20252c" },
    ]);
    expect(ascIcon.props.color).toBe("#20252c");
    expect(ascText.props.style).toStrictEqual([
      { color: "#fff", fontSize: 14, marginLeft: 5 },
      { color: "#20252c" },
    ]);
  });

  it("applies active styles to descending button when sortOrder is 'desc'", () => {
    const { getByText, getByTestId } = render(
      <SortButtons sortOrder="desc" setSortOrder={mockSetSortOrder} />,
    );

    const descButton = getByText("Descendente").parent;
    const descIcon = getByTestId("icon-arrow-down");
    const descText = getByText("Descendente");

    expect(descButton.props.style).toStrictEqual([
      { color: "#fff", fontSize: 14, marginLeft: 5 },
      { color: "#20252c" },
    ]);
    expect(descIcon.props.color).toBe("#20252c");
    expect(descText.props.style).toStrictEqual([
      { color: "#fff", fontSize: 14, marginLeft: 5 },
      { color: "#20252c" },
    ]);
  });

  it("applies inactive styles to ascending button when sortOrder is 'desc'", () => {
    const { getByText, getByTestId } = render(
      <SortButtons sortOrder="desc" setSortOrder={mockSetSortOrder} />,
    );

    const ascButton = getByText("Ascendente").parent;
    const ascIcon = getByTestId("icon-arrow-up");
    const ascText = getByText("Ascendente");

    expect(ascButton.props.style).toStrictEqual([
      { color: "#fff", fontSize: 14, marginLeft: 5 },
      false,
    ]);
    expect(ascIcon.props.color).toBe("#fff");
    expect(ascText.props.style).toStrictEqual([
      { color: "#fff", fontSize: 14, marginLeft: 5 },
      false,
    ]);
  });

  it("matches snapshot", () => {
    const tree = render(
      <SortButtons sortOrder="asc" setSortOrder={mockSetSortOrder} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
