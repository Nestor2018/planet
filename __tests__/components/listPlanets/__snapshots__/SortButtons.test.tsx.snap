// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortButtons Component matches snapshot 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "center",
      "marginVertical": 10,
    }
  }
  testID="sort-buttons"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#3c6fc8",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 5,
        "minWidth": 120,
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="button-asc"
  >
    <view
      color="#20252c"
      name="arrow-up"
      size={18}
      testID="icon-arrow-up"
    />
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontSize": 14,
            "marginLeft": 5,
          },
          {
            "color": "#20252c",
          },
        ]
      }
    >
      Ascendente
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#272c35",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 5,
        "minWidth": 120,
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="button-desc"
  >
    <view
      color="#fff"
      name="arrow-down"
      size={18}
      testID="icon-arrow-down"
    />
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontSize": 14,
            "marginLeft": 5,
          },
          false,
        ]
      }
    >
      Descendente
    </Text>
  </View>
</View>
`;
